### PR TITLE
Custom RPC for Merkle Mountain Range pallet

### PR DESCRIPTION
**Pull request by tomusdrw**
Follow up on #7891.

Initially I thought that it would be possible to use `state_call` RPC to use MMR RuntimeApi and be able to generate proofs.

However, RPC-originated calls DO NOT have access to Offchain DB, hence proof generation is not really possible. I think it's unreasonable to blanket-enable Offchain DB access for all runtime calls over RPC, hence this PR introduces a custom RPC for MMR pallet, which performs RuntimeApi call but with OffchainDB access enabled (`OffchainCall` with default execution extensions enabled).

Unfortunately I had to extract OffchainDB access from `OffchainWorker` context and add yet another execution extension for this to work (I'm sorry @bkchr :))

Companion PR: https://github.com/paritytech/polkadot/pull/2463

### CODE REVIEW

**Comment by tomusdrw** [comment]
@bkchr could you take a look at the externalities extension? Is there any other way how you'd see this being done?

File: bin/node/runtime/src/lib.rs
Comment (adoerr) [inline bin/node/runtime/src/lib.rs:1390]: Perhaps `verify_ephemeral_proof` in order to emphasize, that it is not based on persistent data?

File: frame/merkle-mountain-range/primitives/src/lib.rs
Comment (adoerr) [inline frame/merkle-mountain-range/primitives/src/lib.rs:388]: Shouldn't this throw some kind of error, if `T` doesn't meet the encoding requirements?

**Review by adoerr — APPROVED** [review]
Did focus on the Mmr part, which LGTM

File: frame/merkle-mountain-range/rpc/src/lib.rs
Comment (HCastano) [inline frame/merkle-mountain-range/rpc/src/lib.rs:39]:
```suggestion
/// Retrieved MMR leaf and its proof.
```

File: frame/merkle-mountain-range/rpc/src/lib.rs
Comment (HCastano) [inline frame/merkle-mountain-range/rpc/src/lib.rs:125]: Leftover TODO

File: frame/merkle-mountain-range/primitives/src/lib.rs
Comment (HCastano) [inline frame/merkle-mountain-range/primitives/src/lib.rs:409]: Just curious, why's this okay?

**Review by HCastano — COMMENTED** [review]
The MMR code looks fine, but I don't know anything about the Offchain DB - so I'll leave that for someone else to review

File: bin/node/runtime/src/lib.rs
Comment (tomusdrw) [reply bin/node/runtime/src/lib.rs:1390]: I thought that `stateless` is enough emphasis :) "Ephemeral proof" sounds like a special kind of a proof to me, while here we just don't query the `state`, hence `stateless`.

File: frame/merkle-mountain-range/primitives/src/lib.rs
Comment (tomusdrw) [reply frame/merkle-mountain-range/primitives/src/lib.rs:388]: Hmm, but there isn't really any requirements other than `FullLeaf` implementation. And the encoding is infallible.

So `EncodableOpqaueLeaf` and `OpaqueLeaf` serve a bit different purposes.
Both are meant for leaves that we don't necessarily know the concrete type of (i.e. we know it's a leaf, but we don't have the struct/type to encode it anywhere in the code).

The first one is a wrapper type to send such "unknown leaves" to the RuntimeApi or over RPC.
The second one is for MMR implementation - i.e. it implements `FullLeaf`, but can't implement `Encode/Decode` because of conflicting trait implementations.

`from_leaf` is just a convenience method to convert a known leaf type into its encodable and opaque form - so there is no extra requirements here.

File: frame/merkle-mountain-range/primitives/src/lib.rs
Comment (tomusdrw) [reply frame/merkle-mountain-range/primitives/src/lib.rs:409]: It's not only okay, it's actually required :) By default the `RuntimeApi` will run block initialisation, which triggers `on_initialize` hooks of every modules.
MMR module adds new leaf in `on_initialize`, so the proofs are generated for "the next block" instead of for the current one.

**Comment by bkchr** [comment]
Sorry for the delay, I will review this pr tomorrow.

**Review by kianenigma — COMMENTED** [review]
The runtime portion lgtm, probably you want to wait for one last review from Basti as well.

File: client/api/src/execution_extensions.rs
Comment (bkchr) [inline client/api/src/execution_extensions.rs:99]: Why isn't this just a generic `ExtensionFactory`?

File: client/offchain/src/api.rs
Comment (bkchr) [inline client/offchain/src/api.rs:70]:
```suggestion
		Self { persistent }
```

File: client/offchain/src/api.rs
Comment (bkchr) [inline client/offchain/src/api.rs:111]: For these kind of logging, I can really recommend `tracing::debug!/*` in the future. They provide some nice "structured logging"

General comment, a target would be nice.

File: frame/merkle-mountain-range/primitives/src/lib.rs
Comment (bkchr) [inline frame/merkle-mountain-range/primitives/src/lib.rs:381]:
```suggestion
/// It is different from [`OpaqueLeaf`], because it does implement `Codec`
```

File: frame/merkle-mountain-range/rpc/src/lib.rs
Comment (bkchr) [inline frame/merkle-mountain-range/rpc/src/lib.rs:193]:
```suggestion
		assert_eq!(
			actual,
			r#"{"blockHash":"0x0000000000000000000000000000000000000000000000000000000000000000","leaf":"0x1001020304","proof":"0x010000000000000009000000000000000801010101010101010101010101010101010101010101010101010101010101010202020202020202020202020202020202020202020202020202020202020202"}"#,
	);
```

File: frame/merkle-mountain-range/rpc/src/lib.rs
Comment (bkchr) [inline frame/merkle-mountain-range/rpc/src/lib.rs:218]:
```suggestion
```

File: primitives/runtime/src/offchain/storage.rs
Comment (bkchr) [inline primitives/runtime/src/offchain/storage.rs:140]:
```suggestion
```

We don't need this or?

File: primitives/runtime/src/offchain/storage.rs
Comment (bkchr) [inline primitives/runtime/src/offchain/storage.rs:117]:
```suggestion
```

Same

**Review by bkchr — APPROVED** [review]
Some nitpicks, otherwise looks good.

**Comment by tomusdrw** [comment]
bot merge

**Comment by tomusdrw** [comment]
bot merge force

**Pull request merged**
